src/core/reactor.cc

```diff
@@ -3581,6 +3581,12 @@ smp::get_options_description()
     return opts;
 }
 
+thread_local metrics::impl::metric_implementations metric_impls;
+
+metrics::impl::metric_implementations& metrics::impl::get_metric_implementations() {
+    return metric_impls;
+}
+
```

Comment:
Can you explain the reason for moving this here? Is it to avoid the overhead of a static local?

Reply:
Good question. I initially declared it in `metrics.cc` as that's the most logical place for it, but I ran into issues with the order in which `static thread_local` destructors are called (i.e. reverse order of initialisation). The problem is that the reactor (declared [here](https://github.com/redpanda-data/seastar/blob/0fd17f4cb57365c4a340fbb4435fcc903871512a/src/core/reactor.cc#L3597)) will [de-register](https://github.com/redpanda-data/seastar/blob/master/src/core/metrics.cc#L255) all it's metrics when it goes out of scope (this involves calling `impl::get_local_impl` which acceses `metric_impls`). In order to ensure that `metric_impls` is still alive during this cleanup, I declared it before the [reactor](https://github.com/redpanda-data/seastar/blob/0fd17f4cb57365c4a340fbb4435fcc903871512a/src/core/reactor.cc#L3597).

Another thing that I tried in order to avoid this awkwardly placed declaration was to give each metrics group (`sm::metric_groups`) a `shared_ptr` to the `impl` object. That didn't work as `sm::metric_groups` is movable and can end up being used across shards, which led to the `shared_ptr` being updated from a shard different from where it was created (rightfully, seastar doesn't like that).